packages/autocertifier-server/src/DnsServer.ts

```diff
@@ -0,0 +1,262 @@
+
+        if (parts.length < 3) {
+            // @ts-ignore private field
+            response.header.rcode = NXDOMAIN
+            return send(response)
+        }
+
+        const subdomain = parts[0]
+
+        let retIp = ''
+        if (this.ownHostName === subdomain) {
+            retIp = this.ownIpAddress
+        } else {
+            let subdomainRecord: Subdomain | undefined
+            try {
+                subdomainRecord = await this.db.getSubdomain(subdomain)
+            } catch (e) {
+                logger.error('handleNormalQuery exception')
+            }
+
+            if (!subdomainRecord) {
+                logger.info('handleNormalQuery() not found: ' + name)
+                // @ts-ignore private field
+                response.header.rcode = NXDOMAIN
+                return send(response)
+            }
+            retIp = subdomainRecord.ip
+        }
+
+        response.answers.push({
+            name: mixedCaseName,
+            type: Packet.TYPE.A,
+            class: Packet.CLASS.IN,
+            ttl: 300,
+            address: retIp
+        })
+        send(response)
+    }
+
+    private handleQuery: DnsHandler = async (request: DnsRequest, send: (response: DnsResponse) => void): Promise<void> => {
+
+        const response = Packet.createResponseFromRequest(request)
+        // @ts-ignore private field
+        response.header.aa = 1
+        const question = request.questions[0]
+        if (question === undefined || question.name === undefined) {
+            logger.debug('filtering invalid question')
+            // @ts-ignore private field
+            response.header.rcode = FORMERR
+            return send(response)
+        }
+        const mixedCaseName = question.name
+        const name = mixedCaseName.toLowerCase()
+
+        if (!name.endsWith(this.domainName)) {
+            logger.debug('invalid domain name in query: ' + name)
+            // @ts-ignore private field
+            response.header.rcode = NXDOMAIN
+            return send(response)
+        }
+
+        const parts = mixedCaseName.split('.')
+        const mixedCaseDomainName = parts[parts.length - 2] + '.' + parts[parts.length - 1]
+
+        // TODO: Why DNS2 typing does not expose question.type? The code appears to works and process different
+        // query types correctly.
+        // @ts-ignore private field
+        logger.info(mixedCaseDomainName + ' question type 0x' + Number(question.type).toString(16))
+        // @ts-ignore private field
+        if (question.type === Packet.TYPE.SOA) {
+            return this.handleSOAQuery(mixedCaseDomainName, send, response)
+            // @ts-ignore private field
+        } else if (question.type === Packet.TYPE.NS) {
+            return this.handleNSQuery(mixedCaseDomainName, send, response)
+            // @ts-ignore private field
+        } else if (question.type === Packet.TYPE.TXT) {
+            return this.handleTextQuery(mixedCaseName, send, response)
```

Comment:
Do we use all these queries? (E.g. does someone store TXT entries to our server or query those?)